docs/architecture/adr-108-e2e-abci++.md

```diff
@@ -0,0 +1,296 @@
+		}
+		reqs, err := fetchABCIRequests(t, node.Name)
+		if err != nil {
+			t.Error(fmt.Errorf("Collecting of ABCI requests failed: %w", err))
+		}
+		for i, r := range reqs {
+			isCleanStart := i == 0
+			_, err := checker.Verify(r, isCleanStart)
+			if err != nil {
+				t.Error(fmt.Errorf("ABCI grammar verification failed: %w", err))
+			}
+		}
+	})
+}
+```
+
+Specifically, the test first creates a `GrammarChecker` object. Then for each node in the testnet, it collects all requests 
+logged by this node. Remember here that `fetchABCIRequests()` returns an array of slices(`[]*abci.Request`) where the slice 
+with index 0 corresponds to the node's `CleanStart` execution, and each additional slice corresponds to the `Recovery` 
+execution after a specific crash. Each node must have one `CleanStart` execution and the same number of `Recovery` executions 
+as the number of crashes that happened on this node. If collecting was successful, the test checks whether each execution 
+respects the ABCI++ 
+grammar by calling `checker.Verify()` method. If `Verify` returns an error, the specific execution does not respect the 
+grammar, and the test will fail. 
+
+The tests are executed only if `ABCITestsEnabled` is set to `true`. This is done through the manifest file. Namely, if we 
+want to test whether CometBFT respects ABCI++ grammar, we would need to enable these tests by adding `abci_tests_enabled = 
+true` in the manifest file of a particular testnet (e.g. `networks/ci.toml`). This will automatically activate logging on the 
+application side. 
+
+The `Verify()` method is shown below. 
+```go
+func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {
+	r := g.filterRequests(reqs)
+	// This should not happen in our tests.
+	if len(reqs) == 0 {
+		return false, fmt.Errorf("Execution with no ABCI calls.")
+	}
+	var errors []*Error
+	execution := g.getExecutionString(r)
+	if isCleanStart {
+		errors = g.verifyCleanStart(execution)
+	} else {
+		errors = g.verifyRecovery(execution)
+	}
+	if errors == nil {
+		return true, nil
+	}
+	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
+}
+```
+
+It receives a set of ABCI++ requests and a flag saying whether they represent a `CleanStart` execution or not and does the following things:
+- Filter the requests by calling the method `filterRequests()`. This method will remove all the requests from the set that are not supported by the current version of the grammar. In addition, it will filter the last height by removing all ABCI++ requests after the 
+last `Commit`. The function `fetchABCIRequests()` can be called in the middle of the height. As a result, the last height may be incomplete and 
+classified as invalid, even if that is not the reality. The simple example here is that the last 
+request fetched via `fetchABCIRequests()` is `Decide`; however, `Commit` happens after 
```

Comment:
By `Decide`, did you mean `EndBlock` (or `FinalizeBlock` in `v0.38.x`) ?

Reply:
Decide refers to FinalizeBlock: the terminal name in the grammar is Decide, while the value is FinalizeBlock 
(`decide = %s"<FinalizeBlock>"`). This may need to be changed, but it is like that also in the grammar on main and on v0.38.x branch.

Reply:
I see. In `decide = %s"<FinalizeBlock>"`, technically, `decide` is a non-terminal that gets replaced by terminals "<FinalizeBlock>".
If you think this causes confusion, since we are not planning to backport this project (at least not beyond `v0.38.x`) maybe we can replace `decide` by `finalize`or `finalize-block` or something similar.

Reply:
Yes, we can, just we need to rename it everywhere I guess. In my ADR it is not the problem but not sure where else.

Reply:
I don't think we need to replace all occurrences of the work decide (the concept is clear and understood), just the occurrences of the grammar's `decide` symbol. I quickly checked, and it looks like it's only 4 occurrences.